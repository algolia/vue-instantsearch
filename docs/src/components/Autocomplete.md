---
title: Autocomplete
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/Autocomplete.md
---

Autocomplete is a component which can be used for third-party dropdown components. For one index, or additional indices.

<a class="btn btn-static-theme" href="stories/?selectedKind=Autocomplete">ðŸ•¹ try out live</a>

## Usage

If you want to see what you have access to, simply fill in nothing in 

```html
<ais-autocomplete></ais-autocomplete>
```

## Props

Name | Type | Default | Description | Required
---|---|---|---|---
indices | `Array<{label: String, value: String}>` | `[]` | Additional indices to search in | no
escapeHTML | Boolean | true | Escape raw HTML in the hits | no
classNames | Object | | Override class names | no

## Slots

Name | Scope | Description
---|---|---
default | `{ currentRefinement: string, refine: (value: string) => void, indices: Array<Index> }` | Slot to override the DOM output

`Index` is an object which contains:

Key | Type | Description
---|---|---
index | `String` | The name (value) of this index
label | `String` | The label for this index
hits | `Object[]` | The hits resolved from the index matching the query.
results | `Object` | The full results object from the Algolia API.

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying
DOM structure, have a look at the generated DOM in your browser.

Note that you can pass the prop `class-names`, with an object of class names and their replacement to override this.

Class name | Description
---|---
`ais-Autocomplete` | Container class
