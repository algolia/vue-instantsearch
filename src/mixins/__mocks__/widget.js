let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
