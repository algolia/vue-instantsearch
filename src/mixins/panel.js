import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
