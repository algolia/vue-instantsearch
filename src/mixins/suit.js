import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
