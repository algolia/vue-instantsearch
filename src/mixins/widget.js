const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
