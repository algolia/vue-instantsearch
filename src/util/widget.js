export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
