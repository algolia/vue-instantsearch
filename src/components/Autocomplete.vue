<template>
  <div
    :class="suit()"
    v-if="state"
  >
    <slot
      :refine="state.refine"
      :current-refinement="state.currentRefinement"
      :indices="state.indices"
    >
      <p>This widget doesn't render anything without a filled in default slot.</p>
      <p>query, function to refine and results are provided.</p>
      <pre>refine: Function</pre>
      <pre>currentRefinement: "{{ state.currentRefinement }}"</pre>
      <details>
        <summary><code>indices</code>:</summary>
        <pre>{{ state.indices }}</pre>
      </details>
    </slot>
  </div>
</template>

<script>
import { createWidgetMixin } from '../mixins/widget';
import { connectAutocomplete } from 'instantsearch.js/es/connectors';
import { createSuitMixin } from '../mixins/suit';

export default {
  name: 'AisAutocomplete',
  mixins: [
    createWidgetMixin({ connector: connectAutocomplete }),
    createSuitMixin({ name: 'Autocomplete' }),
  ],
  props: {
    indices: {
      type: Array,
      required: false,
      default: undefined,
    },
    escapeHTML: {
      type: Boolean,
      required: false,
      default: true,
    },
  },
  computed: {
    widgetParams() {
      return {
        indices: this.indices,
        escapeHTML: this.escapeHTML,
      };
    },
  },
};
</script>
