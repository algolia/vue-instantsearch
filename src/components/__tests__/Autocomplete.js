import { mount } from '@vue/test-utils';
import Autocomplete from '../Autocomplete.vue';
import { __setState } from '../../mixins/widget';
jest.mock('../../mixins/widget');

const defaultState = {
  refine: jest.fn(),
  currentRefinement: '',
  indices: [
    {
      index: 'bla',
      label: 'bla bla bla ',
      hits: [{ objectID: 1, name: 'hi' }],
      results: {},
    },
  ],
};

it('renders correctly', () => {
  __setState({
    ...defaultState,
  });
  const wrapper = mount(Autocomplete);

  expect(wrapper.html()).toMatchSnapshot();
});

it('gives the correct props to the default slot', done => {
  __setState({
    ...defaultState,
  });
  mount(Autocomplete, {
    scopedSlots: {
      default(props) {
        expect(props).toEqual(
          expect.objectContaining({
            currentRefinement: defaultState.currentRefinement,
            refine: defaultState.refine,
            indices: defaultState.indices,
          })
        );
        done();
      },
    },
  });
});
