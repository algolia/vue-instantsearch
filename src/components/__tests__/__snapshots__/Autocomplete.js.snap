// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `

<div class="ais-Autocomplete">
  <p>
    This widget doesn't render anything without a filled in default slot.
  </p>
  <p>
    query, function to refine and results are provided.
  </p>
  <pre>
    refine: Function
  </pre>
  <pre>
    currentRefinement: ""
  </pre>
  <details>
    <summary>
      <code>
        indices
      </code>
      :
    </summary>
    <pre>
      [
  {
    "index": "bla",
    "label": "bla bla bla ",
    "hits": [
      {
        "objectID": 1,
        "name": "hi"
      }
    ],
    "results": {}
  }
]
    </pre>
  </details>
</div>

`;
